File: src/main/java/CaptainMarkov/generators/Treknobabble.java
package CaptainMarkov.generators;

import CaptainMarkov.getters.SeriesReader;
import CaptainMarkov.gui.ChainBuilder;
import CaptainMarkov.utils.IterableFile;
import CaptainMarkov.utils.MarkovChain;
import CaptainMarkov.utils.Series;

import java.io.FileNotFoundException;
import java.util.HashSet;

public class Treknobabble implements Generator {
	public MarkovChain chain;
	private HashSet<String> dictionary;

	public Treknobabble(double threshold) {
		chain = new MarkovChain();
		SeriesReader seriesReader = new SeriesReader(Series.TOS);
		loadDictionary();
		while (seriesReader.hasNextSeries()) {
			ChainBuilder.THIS.setLabel("Reading series " + seriesReader.getSeries().toString());
			for (String cur : seriesReader) {
				cur = cur.replaceFirst("[A-Z]+:", "");
				if (getCoherencyRatio(cur) < threshold)
					chain.addWords(cur + "#");
			}
			seriesReader.nextSeries();
		}
	}

	public String generate() {
		return chain.generateSentence();
	}

	/**
	 * Load dictionary from file into HashSet
	 */
	private void loadDictionary() {
		try {
			IterableFile file = new IterableFile("dictionary.txt");
			dictionary = new HashSet<>();
			for (String s : file) {
				dictionary.add(s);
			}
		} catch (FileNotFoundException e) {
			System.out.println("Couldn't find dictionary.txt");
			e.printStackTrace();
		}
	}

	/**
	 * Uses an English dictionary to rate input sentences for how many fake words they contain.
	 * Only accept sentences that are below a (yet to be determined) threshold
	 * Possibly 50%, AKA a threshold of 1
	 *
	 * @param dialogue String
	 */
	private int getCoherencyRatio(String dialogue) {
		int trekWords = 0;
		int realWords = 0;
		String[] words = dialogue.split(" ");
		for (String word : words) {
			if (dictionary.contains(word)) realWords++;
			else trekWords++;
		}
		if(trekWords == 0) return 100;
		return realWords/trekWords;
	}
}

File: CaptainMarkov/src/main/kotlin/captainmarkov/generators/Treknobabble.kt
package captainmarkov.generators

import captainmarkov.getters.SeriesReader
import captainmarkov.gui.ChainBuilder
import captainmarkov.utils.MarkovChain
import captainmarkov.utils.Series
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

class Treknobabble(threshold: Double) : Generator {
    var chain: MarkovChain = MarkovChain()
    private var dictionary: HashSet<String>? = null

    init {
        loadDictionary()
        SeriesReader.readAll()
            .map { it.replaceFirst("[A-Z]+:".toRegex(), "") }
            .filter { getCoherencyRatio(it) < threshold }
            .forEach { chain.addWords("$it#") }
    }

    override fun generate(): String {
        return chain.generateSentence()
    }

    /**
     * Load dictionary from file into HashSet
     */
    private fun loadDictionary() {
        dictionary = HashSet()
        val `in` = this.javaClass.getResourceAsStream("resources/dictionary.txt")
        if (`in` == null) {
            legacyLoadDictionary()
            return
        }
        val scanner = Scanner(`in`)
        var s: String? = null
        while (scanner.hasNextLine()) {
            s = scanner.nextLine()
            dictionary!!.add(s)
        }
    }

    /**
     * Load dictionary with old method to support running program from IDE instead of from JAR
     */
    private fun legacyLoadDictionary() {
        try {
            dictionary!!.addAll(
                Files.readAllLines(Paths.get("src/main/java/captainmarkov/generators/resources/dictionary.txt"))
            )
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    /**
     * Uses an English dictionary to rate input sentences for how many fake words they contain.
     * Only accept sentences that are below a (yet to be determined) threshold
     * Possibly 50%, AKA a threshold of 1
     *
     * @param dialogue String
     */
    private fun getCoherencyRatio(dialogue: String): Int {
        var trekWords = 0
        var realWords = 0
        val words = dialogue.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (word in words) {
            if (dictionary!!.contains(word))
                realWords++
            else
                trekWords++
        }
        return if (trekWords == 0) 100 else realWords / trekWords
    }
}
